From 019da249d0bc23c6a9d60cb08c4a90d360401f4a Mon Sep 17 00:00:00 2001
From: Ricard Wanderlof <polluxsynth@butoba.net>
Date: Sat, 29 Jan 2022 00:33:02 +0100
Subject: [PATCH 1822/1936] Bump version to 3.0.7 to handle portamento legacy

Prior to this version, mono and legato modes always had auto
portamento (i.e. portamento only happened when notes were
played legato), and poly mode never had it (i.e. when portamento
was enabled, every note had portamento regardless of playing
style). Since the introduction of the new auto portamento mode
parameter, where the necessity to play legato is now governed by an
explicit parameter, we need to potentially set this parameter when
loading older files so that the portamento behavior remains the same
as originally intended.
---
 CMakeLists.txt              | 2 +-
 src/Misc/Part.cpp           | 8 ++++++++
 src/Tests/guitar-adnote.xmz | 2 +-
 3 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index ef17dbd3..e678d939 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -3,7 +3,7 @@ set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/")
 project(zynaddsubfx)
 set(VERSION_MAJOR "3")
 set(VERSION_MINOR "0")
-set(VERSION_REVISION "6")
+set(VERSION_REVISION "7")
 
 #Set data directory, if any
 if(DEFINED ZYN_DATADIR)
diff --git a/src/Misc/Part.cpp b/src/Misc/Part.cpp
index c6dbc82b..c62af581 100644
--- a/src/Misc/Part.cpp
+++ b/src/Misc/Part.cpp
@@ -1464,7 +1464,15 @@ void Part::getfromXML(XMLwrapper& xml)
     }
 
     if(xml.enterbranch("CONTROLLER")) {
+        const bool upgrade_3_0_7 = (xml.fileversion() < version_type(3,0,7));
+
         ctl.getfromXML(xml);
+        // Before version 3.0.7, mono and legato modes always had auto
+        // portamento (and poly mode never did, but in that case
+        // portamento.automode will default to off, so no need to do anything
+        // here).
+        if (upgrade_3_0_7 && !Ppolymode)
+            ctl.portamento.automode = 1;
         xml.exitbranch();
     }
 }
diff --git a/src/Tests/guitar-adnote.xmz b/src/Tests/guitar-adnote.xmz
index a3fc06b1..42161578 100644
--- a/src/Tests/guitar-adnote.xmz
+++ b/src/Tests/guitar-adnote.xmz
@@ -2,7 +2,7 @@
 <?xml version="1.0f" encoding="UTF-8"?>
 <!DOCTYPE ZynAddSubFX-data>
 <ZynAddSubFX-data version-major="3" version-minor="0"
-version-revision="6" ZynAddSubFX-author="Nasca Octavian Paul">
+version-revision="7" ZynAddSubFX-author="Nasca Octavian Paul">
 <INFORMATION>
 <par_bool name="PADsynth_used" value="yes" />
 </INFORMATION>
-- 
2.47.0

